Linking issues with 1.17.1 on macOS

**Does this issue affect the google-cloud-cpp project?**

This is a problem with building `google-cloud-cpp` on OSX for the conda-forge distribution.

**What component of `google-cloud-cpp` is this related to?**

`google_cloud_cpp_generator`

**Describe the bug**

When building the currently latest release, we have a linker error with missing symbols in `google_cloud_cpp_generator`. The `google_cloud_cpp_generator` library needs `google::protobuf::compiler::CodeGenerator::~CodeGenerator()` and other functions from this class. Currently the library only links indirectly against `libprotobuf` but not against `libprotoc` where these symbols are located. As clang/lld is more strict in linking than the typical GCC toolchain on Linux, we don't see the issue there. This surfaced during the update of the conda-forge package for `google-cloud-cpp`, see https://github.com/conda-forge/google-cloud-cpp-feedstock/pull/25

```
[519/523] Linking CXX shared library generator/libgoogle_cloud_cpp_generator.dylib
FAILED: generator/libgoogle_cloud_cpp_generator.dylib
: && $BUILD_PREFIX/bin/x86_64-apple-darwin13.4.0-clang++ -march=core2 -mtune=haswell -mssse3 -ftree-vectorize -fPIC -fPIE -fstack-protector-strong -O2 -pipe -stdlib=libc++ -fvisibility-inlines-hidden -std=c++14 -fmessage-length=0 -isystem $PREFIX/include -fdebug-prefix-map=$SRC_DIR=/usr/local/src/conda/google-cloud-cpp-1.17.1 -fdebug-prefix-map=$PREFIX=/usr/local/src/conda-prefix -O3 -DNDEBUG -isysroot /opt/MacOSX10.9.sdk -mmacosx-version-min=10.9 -dynamiclib -Wl,-headerpad_max_install_names -Wl,-pie -Wl,-headerpad_max_install_names -Wl,-dead_strip_dylibs -Wl,-rpath,$PREFIX/lib -L$PREFIX/lib   -framework CoreFoundation -o generator/libgoogle_cloud_cpp_generator.dylib -install_name @rpath/libgoogle_cloud_cpp_generator.dylib generator/CMakeFiles/google_cloud_cpp_generator.dir/generator.cc.o generator/CMakeFiles/google_cloud_cpp_generator.dir/internal/codegen_utils.cc.o generator/CMakeFiles/google_cloud_cpp_generator.dir/internal/descriptor_utils.cc.o generator/CMakeFiles/google_cloud_cpp_generator.dir/internal/predicate_utils.cc.o generator/CMakeFiles/google_cloud_cpp_generator.dir/internal/stub_generator.cc.o  -Wl,-rpath,$SRC_DIR/build_cmake/google/cloud -Wl,-rpath,$SRC_DIR/build_cmake/external/googleapis  google/cloud/libgoogle_cloud_cpp_grpc_utils.1.17.1.dylib  google/cloud/libgoogle_cloud_cpp_common.1.17.1.dylib  external/googleapis/libgoogleapis_cpp_longrunning_operations_protos.1.17.1.dylib  $PREFIX/lib/libgrpc++.1.32.0.dylib  $PREFIX/lib/libprotobuf.dylib  $PREFIX/lib/libgrpc.12.0.0.dylib  $PREFIX/lib/libgpr.12.0.0.dylib  $PREFIX/lib/libaddress_sorting.12.0.0.dylib  $PREFIX/lib/libupb.12.0.0.dylib  $PREFIX/lib/libssl.dylib  $PREFIX/lib/libcrypto.dylib  $PREFIX/lib/libz.dylib  $PREFIX/lib/libcares.2.4.1.dylib  $PREFIX/lib/libre2.dylib  $PREFIX/lib/libabsl_status.dylib  $PREFIX/lib/libabsl_str_format_internal.dylib  $PREFIX/lib/libabsl_cord.dylib  $PREFIX/lib/libabsl_hash.dylib  $PREFIX/lib/libabsl_bad_variant_access.dylib  $PREFIX/lib/libabsl_city.dylib  $PREFIX/lib/libabsl_raw_hash_set.dylib  $PREFIX/lib/libabsl_bad_optional_access.dylib  $PREFIX/lib/libabsl_hashtablez_sampler.dylib  $PREFIX/lib/libabsl_synchronization.dylib  $PREFIX/lib/libabsl_stacktrace.dylib  $PREFIX/lib/libabsl_symbolize.dylib  $PREFIX/lib/libabsl_debugging_internal.dylib  $PREFIX/lib/libabsl_demangle_internal.dylib  $PREFIX/lib/libabsl_graphcycles_internal.dylib  $PREFIX/lib/libabsl_time.dylib  $PREFIX/lib/libabsl_civil_time.dylib  $PREFIX/lib/libabsl_strings.dylib  $PREFIX/lib/libabsl_int128.dylib  $PREFIX/lib/libabsl_strings_internal.dylib  $PREFIX/lib/libabsl_throw_delegate.dylib  $PREFIX/lib/libabsl_time_zone.dylib  $PREFIX/lib/libabsl_malloc_internal.dylib  $PREFIX/lib/libabsl_base.dylib  $PREFIX/lib/libabsl_spinlock_wait.dylib  $PREFIX/lib/libabsl_raw_logging_internal.dylib  $PREFIX/lib/libabsl_log_severity.dylib  $PREFIX/lib/libabsl_dynamic_annotations.dylib  $PREFIX/lib/libabsl_exponential_biased.dylib  external/googleapis/libgoogleapis_cpp_api_client_protos.1.17.1.dylib  external/googleapis/libgoogleapis_cpp_rpc_status_protos.1.17.1.dylib  external/googleapis/libgoogleapis_cpp_rpc_error_details_protos.1.17.1.dylib  external/googleapis/libgoogleapis_cpp_api_annotations_protos.1.17.1.dylib  external/googleapis/libgoogleapis_cpp_api_http_protos.1.17.1.dylib  $PREFIX/lib/libgrpc++.1.32.0.dylib  $PREFIX/lib/libgrpc.12.0.0.dylib  -framework CoreFoundation  $PREFIX/lib/libgpr.12.0.0.dylib  $PREFIX/lib/libaddress_sorting.12.0.0.dylib  $PREFIX/lib/libupb.12.0.0.dylib  -lm  -lpthread  $PREFIX/lib/libssl.dylib  $PREFIX/lib/libcrypto.dylib  $PREFIX/lib/libz.dylib  $PREFIX/lib/libcares.2.4.1.dylib  -lresolv  $PREFIX/lib/libre2.dylib  $PREFIX/lib/libabsl_status.dylib  $PREFIX/lib/libabsl_str_format_internal.dylib  $PREFIX/lib/libabsl_cord.dylib  $PREFIX/lib/libabsl_hash.dylib  $PREFIX/lib/libabsl_bad_variant_access.dylib  $PREFIX/lib/libabsl_city.dylib  $PREFIX/lib/libabsl_raw_hash_set.dylib  $PREFIX/lib/libabsl_bad_optional_access.dylib  $PREFIX/lib/libabsl_hashtablez_sampler.dylib  $PREFIX/lib/libabsl_synchronization.dylib  $PREFIX/lib/libabsl_stacktrace.dylib  $PREFIX/lib/libabsl_symbolize.dylib  $PREFIX/lib/libabsl_debugging_internal.dylib  $PREFIX/lib/libabsl_demangle_internal.dylib  $PREFIX/lib/libabsl_graphcycles_internal.dylib  $PREFIX/lib/libabsl_time.dylib  $PREFIX/lib/libabsl_strings.dylib  $PREFIX/lib/libabsl_strings_internal.dylib  $PREFIX/lib/libabsl_throw_delegate.dylib  $PREFIX/lib/libabsl_int128.dylib  $PREFIX/lib/libabsl_civil_time.dylib  $PREFIX/lib/libabsl_time_zone.dylib  $PREFIX/lib/libabsl_malloc_internal.dylib  $PREFIX/lib/libabsl_base.dylib  $PREFIX/lib/libabsl_raw_logging_internal.dylib  $PREFIX/lib/libabsl_dynamic_annotations.dylib  $PREFIX/lib/libabsl_log_severity.dylib  $PREFIX/lib/libabsl_spinlock_wait.dylib  $PREFIX/lib/libabsl_exponential_biased.dylib  $PREFIX/lib/libprotobuf.dylib  $PREFIX/lib/libabsl_str_format_internal.dylib  $PREFIX/lib/libabsl_strings.dylib  $PREFIX/lib/libabsl_strings_internal.dylib  $PREFIX/lib/libabsl_base.dylib  $PREFIX/lib/libabsl_dynamic_annotations.dylib  $PREFIX/lib/libabsl_spinlock_wait.dylib  $PREFIX/lib/libabsl_int128.dylib  $PREFIX/lib/libabsl_throw_delegate.dylib  $PREFIX/lib/libabsl_raw_logging_internal.dylib  $PREFIX/lib/libabsl_log_severity.dylib && :
ld: warning: -pie being ignored. It is only used when linking a main executable
Undefined symbols for architecture x86_64:
  "google::protobuf::compiler::CodeGenerator::~CodeGenerator()", referenced from:
      google::cloud::generator::Generator::~Generator() in generator.cc.o
      google::cloud::generator::Generator::~Generator() in generator.cc.o
  "google::protobuf::compiler::ParseGeneratorParameter(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::vector<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > >*)", referenced from:
      google::cloud::generator_internal::ProcessCommandLineArgs(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in codegen_utils.cc.o
  "google::protobuf::compiler::CodeGenerator::GenerateAll(std::__1::vector<google::protobuf::FileDescriptor const*, std::__1::allocator<google::protobuf::FileDescriptor const*> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::compiler::GeneratorContext*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >*) const", referenced from:
      vtable for google::cloud::generator::Generator in generator.cc.o
  "typeinfo for google::protobuf::compiler::CodeGenerator", referenced from:
      typeinfo for google::cloud::generator::Generator in generator.cc.o
ld: symbol(s) not found for architecture x86_64
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
```

**To Reproduce**

* clone https://github.com/conda-forge/google-cloud-cpp-feedstock/pull/25 
* Install conda and conda-build
* run `conda build -m .ci_support/osx_64_.yaml recipe`

This should probably also occur without `conda` but e.g. homebrew on macOS.

**Expected behavior**
A clear and concise description of what you expected to happen.

**Operating system:**

`conda-forge` distribution on macOS 10.15 / 10.9

**What compiler and version are you using?**

```
% clang++ -v
clang version 10.0.1
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Users/uwe/miniconda3/conda-bld/google-cloud-cpp_1600089393948/_build_env/bin
```

**What version of `google-cloud-cpp` are you using?**

1.17.1

**Additional context**

I'll submit a PR later.

--- a/generator/CMakeLists.txt
+++ b/generator/CMakeLists.txt
@@ -41,9 +41,12 @@ target_include_directories(google_cloud_cpp_generator
                            PUBLIC ${PROJECT_SOURCE_DIR} ${PROJECT_BINARY_DIR})
 target_link_libraries(
     google_cloud_cpp_generator
-    PUBLIC google_cloud_cpp_grpc_utils google_cloud_cpp_common
+    PUBLIC google_cloud_cpp_grpc_utils
+           google_cloud_cpp_common
            googleapis-c++::api_client_protos
-           googleapis-c++::longrunning_operations_protos absl::str_format)
+           googleapis-c++::longrunning_operations_protos
+           absl::str_format
+           protobuf::libprotoc)
 google_cloud_cpp_add_common_options(google_cloud_cpp_generator)
 target_compile_options(google_cloud_cpp_generator
                        PUBLIC ${GOOGLE_CLOUD_CPP_EXCEPTIONS_FLAG})

